src/models/document/document-content.ts

```diff
@@ -931,68 +956,53 @@ export const DocumentContentModel = types
-        self.sharedModelMap.set(newSharedModelId, pckg as any);
-        // "copy" is already logged above
-      });
-    },
     handleDragCopyTiles(dragTiles: IDragTilesData, rowInfo: IDropRowInfo) {
       const { tiles, sharedModels } = dragTiles;
-      // if this is not dragging any sharedData, proceed as usual
-      if (!sharedModels.length) {
-        this.userCopyTiles(tiles, rowInfo);
-      }
-
-      // For now only handle shared data if tile is new to destination, and is paired 1:1 with a sharedModel
-      else if (sharedModels.length === 1) {
-        const existingProviderIds = self.getSharedModelsByType("SharedDataSet").map((sm) => {
-          return (sm as any).providerId;
-        });
-        const incomingProviderId = (sharedModels[0] as any).providerId;
-        const tileAlreadyCopied = existingProviderIds.includes(incomingProviderId);
 
-        if (tileAlreadyCopied) return console.warn("not handling new drags of same tile");
-        if (tiles.length > 1) return console.warn("not handling multiple tiles in a drag of shared data");
+      // We need to copy the contents of the tiles and the shared models to the destination.
+      // The ids of the tiles will be remapped as part of the copy process, however, so we
+      // need to make sure that the tile references in the shared models are remapped as well.
+      // To accomplish this, we first generate new ids for the tiles, then we update the
+      // shared models to make use of the new ids, then we copy the shared models into the
+      // document, then we copy the tiles, having modified the tile copying code to make
+      // use of the ids we generated at the beginning of the process.
+
+      // generate new tile ids for the dropped/copied tiles
+      const tileIdMap: Record<string, string> = {};
+      const dropTiles = tiles.map(tile => {
+        const newTileId = uniqueId();
+        tileIdMap[tile.tileId] = newTileId;
+        return { ...tile, newTileId };
+      });
 
-        // this is a drag with a sharedModel we can currently handle
-        if (!tileAlreadyCopied && tiles.length === 1) {
-          this.userCopySingleTileWithSharedModel(dragTiles, rowInfo);
+      // map tile ids within shared models and add them to the document
+      sharedModels?.forEach(sharedItem => {
+        // for now, if we already have this shared model, don't copy it again
+        // later, we can support other options (asking the user, auto-merging, etc.)
+        if (self.sharedModelMap.get(sharedItem.modelId)) return;
+
+        const providerId = sharedItem.providerId ? tileIdMap[sharedItem.providerId] : undefined;
+        const tileIds = sharedItem.tileIds.map(tileId => tileIdMap[tileId]);
+        try {
+          const content = JSON.parse(sharedItem.content, (key, value) => {
+            // some shared models, notably SharedDataSet, have an internal providerId that must be mapped
+            return key === "providerId" ? tileIdMap[value] : value;
```

Comment:
Just to be safer we might want to change this to:
```suggestion
            return key === "providerId" && tileIdMap[value] ? tileIdMap[value] : value;
```
Basically only replace the providerId value if it is in the tileIdMap. This way if a providerId key is used for something else it won't be replaced with undefined.

Reply:
I think this has the same effect -- if `tileIdMap[value]` is `undefined`, then `JSON.parse()` omits the property entirely.

Reply:
But in my suggested change if `tileIdMap[value]` is undefined then `value` will be returned instead of `undefined`. Or perhaps the order of operations is different than I expected so I should have written:
`return (key === "providerId" && tileIdMap[value]) ? tileIdMap[value] : value;`

Reply:
My thinking is that if the tile doesn't exist in the tile map, then we shouldn't have any reference to it anywhere, including in the `providerId`, so I think of `undefined` as being the more conservative option.

Reply:
The case I'm thinking of is the shared model that uses `providerId` internally for something else. Its value could be anything. With the current approach `providerId` would have to be a reserved key name at all levels of the shared model content. If we only replaced matching ids, it decreases the chances to almost zero (with the random ids) that any internal `providerId`s would get messed up. 

But I can see your argument too.

Ideally we would have a way to identify tile id's independent of the key. Then we can replace mapped tile ids, and report an error for any tile id that isn't mapped. 

How about this for a compromise: we leave the code as is, but instead of silently returning undefined we return undefined and do a `console.error` if the id doesn't match. This will help us know if we've got a case where this is a problem.

Reply:
I see your point as well and don't feel too strongly about it either way, so I'm fine with adding a `console.error`.

Reply:
After sleeping on this, I realized that the variables shared model setup with its references from tiles to models within the shared model are going to not work with this approach. So I started looking at generic MST cloning options. With certain conventions, we can use MST's "schema" to identify and update these ids and their references. I'll try to write something up about it.